File: pa1/List.java
// List.java
// An integer List ADT
// elbharri
// pa1
// 1-16-15

class List{

   private class Node{
      // Fields
      int data;
      Node next;
      Node prev; 
      // Constructor
      Node(int data) { this.data = data; next = null; prev = null;}
      // toString:  Overides Object's toString method.
      public String toString() { return String.valueOf(data); }
   }

   // Fields
   private Node front;
   private Node back;
   private Node cursor;
   private int length;
   private int index = -1; 

   // Constructor
   List() { front = back = null; length = 0; }


   // Access Functions //////////////////////////////////////////////////////////////////

   // Returns number of elements in this list.
   int length(){
      return length;

   }
   // Returns the index of the cursor element in this list, or
   //returns -1 if the cursor element is undefined.
   int getIndex(){
      return index;

   }

   // isEmpty(): returns true if this is an empty List, false otherwise
   boolean isEmpty() { return length==0; }

   // Returns front element in this list. Pre: length()>0
   int front(){
       if( this.isEmpty() ){
         throw new RuntimeException("List Error: front() called on empty List");
      }
      return front.data;

   }
   // Returns back element in this list. Pre: length()>0
   int back(){
      if( this.isEmpty() ){
         throw new RuntimeException("List Error: back() called on empty List");
      }
      return back.data;
   }

   // Returns cursor element in this list. Pre: getIndex () >=0
   int getElement(){
      if( this.isEmpty() ){
         throw new RuntimeException("List Error: getElement called on empty List");
      }
      return cursor.data;
   }

   // Returns true if this List and L are the same integer
   // sequence. The cursor is ignored in both lists.
   boolean equals(List L){
      boolean flag  = true;
      Node N = this.front;
      Node M = L.front;

      if( this.length==L.length ){
         while( flag && N!=null){
            flag = (N.data==M.data);
            N = N.next;
            M = M.next;
         }
         return flag;
      }else{
         return false;
      }
   }

   // Manipulation procedures-----------------------------------
   // Re-sets this List to the empty state.
   void clear(){
      front = back = null;
      cursor = null; 
      length = 0;
      index = -1;
   }

   // If 0<=i<=length()-1, moves the cursor to the element
   // at index i, otherwise the cursor becomes undefined.
   void moveTo(int i){
      int n;
      if(this.isEmpty() ){
         throw new RuntimeException("List Error: moveTo called on empty List");
      }else if(i <= length-1 && i >= 0){
         cursor = front;
         for(n = 1; n<=i; n++){
            cursor = cursor.next;
         }
         index = i; 
         
      }

   }
   
   // If 0<getIndex()<=length()-1, moves the cursor one step toward the
   // front of the list. If getIndex()==0, cursor becomes undefined.
   // If getIndex()==-1, cursor remains undefined. This operation is
   // equivalent to moveTo(getIndex()-1).
   void movePrev(){
      if(index == 0 || index == -1 ){
         cursor = null;
         index = -1;
      }else if(index > 0 && index <= length-1){
         cursor = cursor.prev;
         index--;
      }

   }
   
   // If 0<=getIndex()<length()-1, moves the cursor one step toward the
   // back of the list. If getIndex()==length()-1, cursor becomes
   // undefined. If getIndex()==-1, cursor remains undefined. This
   // operation is equivalent to moveTo(getIndex()+1).
   void moveNext(){
      if(index == length-1){
         cursor = null;
         index = -1;
      }
      else if(index == -1){
         cursor = null;
         index = -1;
      }
      else if(index >= 0 && index < length-1){
         cursor = cursor.next;
         index++;
      }

   }

   // Inserts new element before front element in this List.
   void prepend(int data){
      Node node = new Node(data);
      if( this.isEmpty() ){front = back = node; }
      else{
         front.prev = node;
         node.next = front;
         front = node; 
         index++;
      }
      length++;

   }

   // Inserts new element after back element in this List.
   void append(int data){
      Node node = new Node(data);
      if( this.isEmpty() ) { front = back = node; }
      else { 
         back.next = node; 
         node.prev = back;
         back = node; 
      }
      length++;
   }

   // Inserts new element before cursor element in this List. Pre: length()>0, getIndex()>=0
   void insertBefore(int data){
      Node node = new Node(data);
      if( this.isEmpty() ){front = back = node; }
      else{
         node.next = cursor;
         if(cursor.prev != null){
         node.prev = cursor.prev;
         (cursor.prev).next = node;
         }
         cursor.prev = node;
         if (node.prev == null)
      	    front = node;
      }
      index++;
      length++;
   }

   // Inserts new element after cursor element in this
   // List. Pre: length()>0, getIndex()>=0
   void insertAfter(int data){
      Node node = new Node(data);
      if(this.isEmpty() ){
         front = back = node;
      	 index++;
      }else{
         (cursor.next).prev = node;
      	 node.next = cursor.next;
      	 node.prev = cursor;
      	 cursor.next = node;
      	 if(node.next == null)
      	   back = node; 

      }
      length++;
   }
   
   // Deletes the front element in this List. Pre: length()>0
   void deleteFront(){
      if(this.isEmpty()){
         throw new RuntimeException("List Error: Delete front() called on empty List");
      }
      if(this.length>1) {
         front.next.prev = null;
      	 front = front.next;
      } else {
      front = back = null;
      }
      length--;
      index--;
   }

   // Deletes the back element in this List. Pre: length()>0
   void deleteBack(){
      if(this.isEmpty()){
         throw new RuntimeException("List Error: DeleteBack called on empty List");
      }
      if(this.length>1){
         back.prev.next = null;
      	 back = back.prev; 
         length--;     
      }
   }
   
   // Deletes cursor element in this List. Cursor is undefined after this
   // operation. Pre: length()>0, getIndex()>=0
   void delete(){
      if(this.isEmpty()){
         throw new RuntimeException("List Error: delete called on an empty list");
      }else if(index<0){
         throw new RuntimeException("List Error: delete called on a null cursor");
      }else if(index>=0){
         cursor.prev.next = cursor.next;
         cursor.next.prev = cursor.prev;
         cursor = null;
         length--;
         index = -1;
      }   
   }
   //Other methods------------------------------------------

   // Overides Object's toString method. Creates a string
   // consisting of a space separated sequence of integers 
   // with front on the left and back on the right. The
   // cursor is ignored.
   public String toString(){
      String str = "";
      for(Node N=front; N!=null; N=N.next){
         str += N.toString() + " ";
      }
      return str;
   }
   
   // copy(): returns a new List identical to this one.
   // Returns a new list representing the same integer sequence as this
   // list. The cursor in the new list is undefined, regardless of the
   // state of the cursor in this List. This List is unchanged.
   List copy(){
      List L = new List();
      Node N = this.front;
      while( N!=null ){
         L.append(N.data);
         N = N.next;
      }
      return L;
   }
 

}

File: pa2/List.c
//-----------------------------------------------------------------------------
// List.c
//CMPS101
//Elizabeth Harris
//elbharri
//pa2
//1-23-15
// Implementation file for List ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// structs --------------------------------------------------------------------

// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;


// Constructors-Destructors ---------------------------------------------------

// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->prev=NULL;
   return(N);
}

// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = NULL;
   L->cursor = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}


// freeList()
// Frees all heap memory associated with List *pQ, and sets *pQ to NULL.S
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) {
      while( !isEmpty(*pL) ) {
         deleteFront(*pL);
      }
      free(*pL);
      *pL = NULL;
   }
}

// Access functions -----------------------------------------------------------


// getLength()
// Returns the length of L.
int length(List L){
   if( L==NULL ){
      printf("List Error: calling getLength() on NULL List reference\n");
      exit(1);
   }
   return(L->length);
}

// isEmpty()
// Returns true (1) if L is empty, otherwise returns false (0)
int isEmpty(List L){
   if( L==NULL ){
      printf("List Error: calling isEmpty() on NULL List reference\n");
      exit(1);
   }
   return(L->length==0);
}

//getIndex(List L)
//returns index
int getIndex(List L){
   if( L==NULL ){
      printf("List Error: calling getLength() on NULL List reference\n");
      exit(1);
   }return(L->index);

}

// front()
// Returns the value at the front of L.
// // Pre: !isEmpty(L)
int front(List L){
   if( L==NULL ){
      printf("List Error: calling getFront() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling getFront() on an empty List\n");
      exit(1);
   }
   return(L->front->data);
}

//int back(List L)
//returns back element in list
int back(List L){
    if( L==NULL ){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling back() on an empty List\n");
      exit(1);
   }
   return(L->back->data);

}

//getElement()
//returns element at cursor
int getElement(List L){
   if( L==NULL ){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling back() on an empty List\n");
      exit(1);
   }
   return(L->cursor->data);
}

// equals()
// returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B){
   int flag = 1;
   Node N = NULL;
   Node M = NULL;

   if( A==NULL || B==NULL ){
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   N = A->front;
   M = B->front;
   if( A->length != B->length ) { return 0; }
   while( flag && N!=NULL){
      flag = (N->data==M->data);
      N = N->next;
      M = M->next;
   }
   return flag;
}
// Manipulation procedures ----------------------------------------------------

//clear()
//clears list to null state
void clear(List L){
   int i;
   L->front = L->back = NULL;
   L->cursor = NULL;
   L->length = 0;
   L->index = -1;
  // for(i = 0; i<L->length-1;)
}

//moveto()
//moves cursor to index specified
void moveTo(List L, int i){
   int n;
   if( L==NULL ){
      printf("List Error: calling moveTo() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling moveTo() on an empty List\n");
      exit(1);
   }else if(i <= L->length-1 && i >= 0){
      L->cursor = L->front;
      for(n = 1; n <= i; n++){
         L->cursor = L->cursor->next;
      }
      L->index = i;
   }

}

//moves cursor to previous node
void movePrev(List L){
   if( L==NULL ){
      printf("List Error: calling moveTo() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling moveTo() on an empty List\n");
      exit(1);
   }else if(L->index == -1 || L->index == 0){
         L->cursor = NULL;
         L->index = -1;
      }
      else if(L->index > 0 && L->index <= L->length-1){
         L->cursor = L->cursor->prev;
         L->index--;
      }

}

//moveNext(List L)
//moves cursor to next node
void moveNext(List L){
   if( L==NULL ){
      printf("List Error: calling moveTo() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling moveTo() on an empty List\n");
      exit(1);
   }else if(L->index == -1 || L->index == (L->length)-1){
         L->cursor = NULL;
         L->index = -1;
   }
   else if(L->index >= 0 && L->index <= (L->length)-1){
         L->cursor = L->cursor->next;
         L->index++;
      }

}

//prepend(List L, int data)
//inserts new data element at the front of List L
void prepend(List L, int data){
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling EnList() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ) { L->front = L->back = N; }
   else { 
      L->front->prev = N;
      N->next = L->front; 
      L->front = N;
      L->index++;
   }
   L->length++;
}

// EnList()
// Places new data element at the end of L
void append(List L, int data){
   Node N = newNode(data);

   if( L==NULL ){
      printf("List Error: calling EnList() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ) { L->front = L->back = N; }
   else {
      L->back->next = N;
      N->prev = L->back;
      L->back = N;
   }
   L->length++;
}

//insertBefore(List L, int data);
//Places new data element before cursor
void insertBefore(List L, int data){
   Node N = newNode(data);
   if( L==NULL ){
      printf("List Error: calling EnList() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ) { L->front = L->back = N; }
   else{
      N->next = L->cursor;
      if(L->cursor->prev != NULL){
         N->prev = L->cursor->prev;
         (L->cursor->prev->next) = N;
      }
      L->cursor->prev = N;
      if(N->prev == NULL)
         L->front = N;
   }
   L->index++;
   L->length++;
}

//insertAfter(List L, int data);
//places new data element after cursor
void insertAfter(List L, int data){
   Node N = newNode(data);
   if( L==NULL ){
      printf("List Error: calling EnList() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      L->front = L->back = N;
      L->index++;
   }else{
      L->cursor->next->prev = N;
      N->next = L->cursor->next;
      N->prev = L->cursor;
      L->cursor->next = N;
      if(N->next == NULL)
         L->back = N;
   }
   L->length++;
}

// DeleteFront()
// Deletes element at front of L
// Pre: !isEmpty(Q)
void deleteFront(List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling Delete front() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling Delete front on an empty List\n");
      exit(1);
   }
   if(L->index == 0){
      L->index = -1;
      L->cursor = NULL;
   }
   N = L->front;
   if( length(L) > 1 ) { L->front = L->front->next; }
   
   L->length--;
   //N = NULL;
   freeNode(&N);
}

//DeleteBack
//deletes element at the back of the list
void deleteBack(List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling DeList() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling DeList on an empty List\n");
      exit(1);
   }
   N = L->back;
   if( L->length > 1 ) {L->back = L->back->prev; }
   else {
      L->front = L->back = NULL;
      //freeNode(L->front);
   }
   L->length--;
   N = NULL;
   freeNode(&N);
}

//delete
//delete element at cursor
void delete(List L){
   Node N = NULL;
   if( L==NULL ){
      printf("List Error: calling DeList() on NULL List reference\n");
      exit(1);
   }
   if( isEmpty(L) ){
      printf("List Error: calling DeList on an empty List\n");
      exit(1);
   }else if(getIndex(L) == -1){
      printf("List Error: calling delete on a null cursor");
      exit(1);
   }else if(getIndex(L) >= 0){
      if(L->cursor->prev == NULL){
         L->front = L->cursor->next;
      }
      else if(L->cursor->next == NULL){
         L->back = L->cursor->prev;
      }
      N = L->cursor;
      L->cursor->prev->next = L->cursor->next;
      L->cursor->next->prev = L->cursor->prev;
      L->cursor = NULL;
      L->length--;
      L->index = -1;
      freeNode(&N);
   }


}

// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L){
   Node N = NULL;

   if( L==NULL ){
      printf("List Error: calling printList() on NULL List reference\n");
      exit(1);
   }
   for(N = L->front; N != NULL; N = N->next){
      printf("%d ", N->data);
   }
   printf("\n");
}

   // copy(): returns a new List identical to this one.
   List copyList(List L){
      List A = newList();
      Node N = L->front;

      while( N!=NULL ){
         append(A, N->data);
         N = N->next;
      }
      return A;
   }

 
